// @angular-mdc/web/esm5/typography.es5.js
/**
 * @license
 * Copyright (c) 2018 Dominic Carretto
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/trimox/angular-mdc-web/blob/master/LICENSE
 */
import { Directive, NgModule } from '@angular/core';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var MdcTypography = /** @class */ (function () {
    function MdcTypography() {
    }
    MdcTypography.decorators = [
        { type: Directive, args: [{
                    selector: '[mdc-typography], [mdcTypography]',
                    host: { 'class': 'mdc-typography' }
                },] },
    ];
    return MdcTypography;
}());
var MdcHeadline1 = /** @class */ (function () {
    function MdcHeadline1() {
    }
    MdcHeadline1.decorators = [
        { type: Directive, args: [{
                    selector: '[mdcHeadline1]',
                    host: { 'class': 'mdc-typography--headline1' }
                },] },
    ];
    return MdcHeadline1;
}());
var MdcHeadline2 = /** @class */ (function () {
    function MdcHeadline2() {
    }
    MdcHeadline2.decorators = [
        { type: Directive, args: [{
                    selector: '[mdcHeadline2]',
                    host: { 'class': 'mdc-typography--headline2' }
                },] },
    ];
    return MdcHeadline2;
}());
var MdcHeadline3 = /** @class */ (function () {
    function MdcHeadline3() {
    }
    MdcHeadline3.decorators = [
        { type: Directive, args: [{
                    selector: '[mdcHeadline3]',
                    host: { 'class': 'mdc-typography--headline3' }
                },] },
    ];
    return MdcHeadline3;
}());
var MdcHeadline4 = /** @class */ (function () {
    function MdcHeadline4() {
    }
    MdcHeadline4.decorators = [
        { type: Directive, args: [{
                    selector: '[mdcHeadline4]',
                    host: { 'class': 'mdc-typography--headline4' }
                },] },
    ];
    return MdcHeadline4;
}());
var MdcHeadline5 = /** @class */ (function () {
    function MdcHeadline5() {
    }
    MdcHeadline5.decorators = [
        { type: Directive, args: [{
                    selector: '[mdcHeadline5]',
                    host: { 'class': 'mdc-typography--headline5' }
                },] },
    ];
    return MdcHeadline5;
}());
var MdcHeadline6 = /** @class */ (function () {
    function MdcHeadline6() {
    }
    MdcHeadline6.decorators = [
        { type: Directive, args: [{
                    selector: '[mdcHeadline6]',
                    host: { 'class': 'mdc-typography--headline6' }
                },] },
    ];
    return MdcHeadline6;
}());
var MdcSubtitle1 = /** @class */ (function () {
    function MdcSubtitle1() {
    }
    MdcSubtitle1.decorators = [
        { type: Directive, args: [{
                    selector: '[mdcSubtitle1]',
                    host: { 'class': 'mdc-typography--subtitle1' }
                },] },
    ];
    return MdcSubtitle1;
}());
var MdcSubtitle2 = /** @class */ (function () {
    function MdcSubtitle2() {
    }
    MdcSubtitle2.decorators = [
        { type: Directive, args: [{
                    selector: '[mdcSubtitle2]',
                    host: { 'class': 'mdc-typography--subtitle2' }
                },] },
    ];
    return MdcSubtitle2;
}());
var MdcBody2 = /** @class */ (function () {
    function MdcBody2() {
    }
    MdcBody2.decorators = [
        { type: Directive, args: [{
                    selector: '[mdcBody2]',
                    host: { 'class': 'mdc-typography--body2' }
                },] },
    ];
    return MdcBody2;
}());
var MdcBody1 = /** @class */ (function () {
    function MdcBody1() {
    }
    MdcBody1.decorators = [
        { type: Directive, args: [{
                    selector: '[mdcBody1]',
                    host: { 'class': 'mdc-typography--body1' }
                },] },
    ];
    return MdcBody1;
}());
var MdcCaption = /** @class */ (function () {
    function MdcCaption() {
    }
    MdcCaption.decorators = [
        { type: Directive, args: [{
                    selector: '[mdcCaption]',
                    host: { 'class': 'mdc-typography--caption' }
                },] },
    ];
    return MdcCaption;
}());
var MdcTypographyButton = /** @class */ (function () {
    function MdcTypographyButton() {
    }
    MdcTypographyButton.decorators = [
        { type: Directive, args: [{
                    selector: '[mdcButton]',
                    host: { 'class': 'mdc-typography--button' }
                },] },
    ];
    return MdcTypographyButton;
}());
var MdcOverline = /** @class */ (function () {
    function MdcOverline() {
    }
    MdcOverline.decorators = [
        { type: Directive, args: [{
                    selector: '[mdcOverline]',
                    host: { 'class': 'mdc-typography--overline' }
                },] },
    ];
    return MdcOverline;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var TYPOGRAPHY_DECLARATIONS = [
    MdcTypography,
    MdcBody1,
    MdcBody2,
    MdcTypographyButton,
    MdcCaption,
    MdcHeadline1,
    MdcHeadline2,
    MdcHeadline3,
    MdcHeadline4,
    MdcHeadline5,
    MdcHeadline6,
    MdcOverline,
    MdcSubtitle1,
    MdcSubtitle2
];
var MdcTypographyModule = /** @class */ (function () {
    function MdcTypographyModule() {
    }
    MdcTypographyModule.decorators = [
        { type: NgModule, args: [{
                    exports: TYPOGRAPHY_DECLARATIONS,
                    declarations: TYPOGRAPHY_DECLARATIONS
                },] },
    ];
    return MdcTypographyModule;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { MdcTypographyModule, MdcTypography, MdcHeadline1, MdcHeadline2, MdcHeadline3, MdcHeadline4, MdcHeadline5, MdcHeadline6, MdcSubtitle1, MdcSubtitle2, MdcBody2, MdcBody1, MdcCaption, MdcTypographyButton, MdcOverline };
