// @angular-mdc/web/esm2015/typography.js
/**
 * @license
 * Copyright (c) 2018 Dominic Carretto
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/trimox/angular-mdc-web/blob/master/LICENSE
 */
import { Directive, NgModule } from '@angular/core';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class MdcTypography {
}
MdcTypography.decorators = [
    { type: Directive, args: [{
                selector: '[mdc-typography], [mdcTypography]',
                host: { 'class': 'mdc-typography' }
            },] },
];
class MdcHeadline1 {
}
MdcHeadline1.decorators = [
    { type: Directive, args: [{
                selector: '[mdcHeadline1]',
                host: { 'class': 'mdc-typography--headline1' }
            },] },
];
class MdcHeadline2 {
}
MdcHeadline2.decorators = [
    { type: Directive, args: [{
                selector: '[mdcHeadline2]',
                host: { 'class': 'mdc-typography--headline2' }
            },] },
];
class MdcHeadline3 {
}
MdcHeadline3.decorators = [
    { type: Directive, args: [{
                selector: '[mdcHeadline3]',
                host: { 'class': 'mdc-typography--headline3' }
            },] },
];
class MdcHeadline4 {
}
MdcHeadline4.decorators = [
    { type: Directive, args: [{
                selector: '[mdcHeadline4]',
                host: { 'class': 'mdc-typography--headline4' }
            },] },
];
class MdcHeadline5 {
}
MdcHeadline5.decorators = [
    { type: Directive, args: [{
                selector: '[mdcHeadline5]',
                host: { 'class': 'mdc-typography--headline5' }
            },] },
];
class MdcHeadline6 {
}
MdcHeadline6.decorators = [
    { type: Directive, args: [{
                selector: '[mdcHeadline6]',
                host: { 'class': 'mdc-typography--headline6' }
            },] },
];
class MdcSubtitle1 {
}
MdcSubtitle1.decorators = [
    { type: Directive, args: [{
                selector: '[mdcSubtitle1]',
                host: { 'class': 'mdc-typography--subtitle1' }
            },] },
];
class MdcSubtitle2 {
}
MdcSubtitle2.decorators = [
    { type: Directive, args: [{
                selector: '[mdcSubtitle2]',
                host: { 'class': 'mdc-typography--subtitle2' }
            },] },
];
class MdcBody2 {
}
MdcBody2.decorators = [
    { type: Directive, args: [{
                selector: '[mdcBody2]',
                host: { 'class': 'mdc-typography--body2' }
            },] },
];
class MdcBody1 {
}
MdcBody1.decorators = [
    { type: Directive, args: [{
                selector: '[mdcBody1]',
                host: { 'class': 'mdc-typography--body1' }
            },] },
];
class MdcCaption {
}
MdcCaption.decorators = [
    { type: Directive, args: [{
                selector: '[mdcCaption]',
                host: { 'class': 'mdc-typography--caption' }
            },] },
];
class MdcTypographyButton {
}
MdcTypographyButton.decorators = [
    { type: Directive, args: [{
                selector: '[mdcButton]',
                host: { 'class': 'mdc-typography--button' }
            },] },
];
class MdcOverline {
}
MdcOverline.decorators = [
    { type: Directive, args: [{
                selector: '[mdcOverline]',
                host: { 'class': 'mdc-typography--overline' }
            },] },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const TYPOGRAPHY_DECLARATIONS = [
    MdcTypography,
    MdcBody1,
    MdcBody2,
    MdcTypographyButton,
    MdcCaption,
    MdcHeadline1,
    MdcHeadline2,
    MdcHeadline3,
    MdcHeadline4,
    MdcHeadline5,
    MdcHeadline6,
    MdcOverline,
    MdcSubtitle1,
    MdcSubtitle2
];
class MdcTypographyModule {
}
MdcTypographyModule.decorators = [
    { type: NgModule, args: [{
                exports: TYPOGRAPHY_DECLARATIONS,
                declarations: TYPOGRAPHY_DECLARATIONS
            },] },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { MdcTypographyModule, MdcTypography, MdcHeadline1, MdcHeadline2, MdcHeadline3, MdcHeadline4, MdcHeadline5, MdcHeadline6, MdcSubtitle1, MdcSubtitle2, MdcBody2, MdcBody1, MdcCaption, MdcTypographyButton, MdcOverline };
