// dialog-service/esm2015/lib/mat/Form.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, Inject } from '@angular/core';
import { DialogService } from '../DialogService';
import { BaseFormComponent } from '../BaseFormComponent';
import { MAT_DIALOG_DATA, MatDialogRef } from '@angular/material/dialog';
// @dynamic
export class FormComponent extends BaseFormComponent {
    /**
     * @param {?} cd
     * @param {?} dialogRef
     * @param {?} dialogService
     * @param {?} data
     */
    constructor(cd, dialogRef, dialogService, data) {
        super(cd, dialogService, dialogRef, data.cancelMessage, data.fields, data.rawValue, data.valueChanges, data.formCreated);
        this.data = data;
    }
    /**
     * @return {?}
     */
    get attrFlexCell() {
        /** @type {?} */
        const flexCell = this.data.layout && this.data.layout.flexCell
        // must return null for Angular to remove the attribute
        ;
        // must return null for Angular to remove the attribute
        return flexCell ? true : null;
    }
    /**
     * @return {?}
     */
    get attrGutter() {
        /** @type {?} */
        const gutter = this.data.layout && this.data.layout.gutter
        // must return null for Angular to remove the attribute
        ;
        // must return null for Angular to remove the attribute
        return gutter ? true : null;
    }
    /**
     * @return {?}
     */
    get attrGrowItems() {
        /** @type {?} */
        const growItems = this.data.layout && this.data.layout.growItems
        // must return null for Angular to remove the attribute
        ;
        // must return null for Angular to remove the attribute
        return growItems ? true : null;
    }
    /**
     * @return {?}
     */
    get attrDefaultCell6() {
        /** @type {?} */
        const flexCell = this.data.layout && this.data.layout.flexCell
        // must return null for Angular to remove the attribute
        ;
        // must return null for Angular to remove the attribute
        return flexCell === 6 ? true : null;
    }
    /**
     * @return {?}
     */
    get attrDefaultCell12() {
        /** @type {?} */
        const flexCell = this.data.layout && this.data.layout.flexCell
        // must return null for Angular to remove the attribute
        ;
        // must return null for Angular to remove the attribute
        return flexCell === 12 ? true : null;
    }
    /**
     * @return {?}
     */
    get attrDebug() {
        /** @type {?} */
        const debug = this.data.layout && this.data.layout.debug
        // must return null for Angular to remove the attribute
        ;
        // must return null for Angular to remove the attribute
        return debug ? true : null;
    }
}
FormComponent.decorators = [
    { type: Component, args: [{
                template: "<h1 mat-dialog-title>{{data.title}}</h1>\n<div mat-dialog-content>\n    <div *ngIf=\"data.content\" [innerHtml]=\"data.content\" class=\"content\"></div>\n    <div>\n        <form [formGroup]=\"form\"\n              [attr.flex-cell]=\"attrFlexCell\"\n              [attr.gutter]=\"attrGutter\"\n              [attr.grow-items]=\"attrGrowItems\"\n              [attr.default-cell-6]=\"attrDefaultCell6\"\n              [attr.default-cell-12]=\"attrDefaultCell12\"\n              [attr.debug]=\"attrDebug\"\n              autocomplete=\"off\">\n            <quick-form-field [field]=\"field\" [form]=\"form\"\n                              *ngFor=\"let field of formFields\"></quick-form-field>\n        </form>\n    </div>\n</div>\n<div mat-dialog-actions>\n    <button mat-button color=\"primary\" type=\"button\"\n            (click)=\"doSubmit()\">{{data.submitButton}}</button>\n    <button mat-button type=\"button\" (click)=\"doClose()\">{{data.cancelButton}}</button>\n</div>\n",
                changeDetection: ChangeDetectionStrategy.OnPush
            }] }
];
/** @nocollapse */
FormComponent.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: MatDialogRef },
    { type: DialogService },
    { type: undefined, decorators: [{ type: Inject, args: [MAT_DIALOG_DATA,] }] }
];
if (false) {
    /** @type {?} */
    FormComponent.prototype.data;
}
