// dialog-service/esm5/lib/BaseFormComponent.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { debounceTime, filter, startWith, take, tap } from 'rxjs/operators';
import { assert } from './util/assert';
import { QuickForm } from 'ng-quick-form';
import { markAllAsDirtyAndTouched } from './util/markAllAsDirtyAndTouched';
/**
 * Manage form creation and submission.
 */
var /**
 * Manage form creation and submission.
 */
BaseFormComponent = /** @class */ (function () {
    function BaseFormComponent(cd, dialogService, closeable, cancelMessage, formFields, rawValue, valueChanges, formCreated) {
        var _this = this;
        this.cd = cd;
        this.dialogService = dialogService;
        this.closeable = closeable;
        this.cancelMessage = cancelMessage;
        this.formFields = formFields;
        this.rawValue = rawValue;
        this.valueChanges = valueChanges;
        this.formCreated = formCreated;
        this.submitted = false;
        this.form = QuickForm.makeForm(this.formFields);
        if (formCreated) {
            formCreated(this.form, this.cd);
        }
        if (this.valueChanges) {
            this.formSubscription = this.form.valueChanges.pipe(debounceTime(100), tap((/**
             * @param {?} value
             * @return {?}
             */
            function (value) { return (/** @type {?} */ (_this.valueChanges)).next({
                value: value, form: _this.form, cd: cd
            }); }))).subscribe();
        }
    }
    /**
     * @return {?}
     */
    BaseFormComponent.prototype.doClose = /**
     * @return {?}
     */
    function () {
        var _this = this;
        // close form dialog, return false as result
        this.dialogService.withConfirm(this.cancelMessage).pipe(filter((/**
         * @param {?} confirm
         * @return {?}
         */
        function (confirm) { return confirm; })), tap((/**
         * @return {?}
         */
        function () { return _this.closeable.close(false); }))).subscribe();
    };
    /**
     * @return {?}
     */
    BaseFormComponent.prototype.doSubmit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.form.status === 'PENDING') {
            // if form is pending validation, let's wait...
            this.dialogService.withProgress(this.form.statusChanges.pipe(startWith(this.form.status), filter((/**
             * @param {?} status
             * @return {?}
             */
            function (status) { return status !== 'PENDING'; })), take(1), tap((/**
             * @return {?}
             */
            function () { return _this.doNonPendingSubmit(); }))));
        }
        else {
            this.doNonPendingSubmit();
        }
    };
    /**
     * @return {?}
     */
    BaseFormComponent.prototype.doNonPendingSubmit = /**
     * @return {?}
     */
    function () {
        assert(this.form.status !== 'PENDING', 'Form status should not be pending');
        markAllAsDirtyAndTouched(this.form);
        this.submitted = true; // attempted to submit (affects validation messages)
        this.cd.markForCheck();
        if (!this.form.valid) {
            return;
        }
        // close form dialog, return form value object as result
        this.closeable.close(QuickForm.preProcessFormValues(this.rawValue ? this.form.getRawValue() : this.form.value));
    };
    /**
     * @return {?}
     */
    BaseFormComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this.formSubscription) {
            this.formSubscription.unsubscribe();
        }
    };
    return BaseFormComponent;
}());
/**
 * Manage form creation and submission.
 */
export { BaseFormComponent };
if (false) {
    /** @type {?} */
    BaseFormComponent.prototype.submitted;
    /** @type {?} */
    BaseFormComponent.prototype.form;
    /** @type {?} */
    BaseFormComponent.prototype.formSubscription;
    /**
     * @type {?}
     * @private
     */
    BaseFormComponent.prototype.cd;
    /**
     * @type {?}
     * @private
     */
    BaseFormComponent.prototype.dialogService;
    /**
     * @type {?}
     * @private
     */
    BaseFormComponent.prototype.closeable;
    /**
     * @type {?}
     * @private
     */
    BaseFormComponent.prototype.cancelMessage;
    /** @type {?} */
    BaseFormComponent.prototype.formFields;
    /**
     * @type {?}
     * @private
     */
    BaseFormComponent.prototype.rawValue;
    /**
     * @type {?}
     * @private
     */
    BaseFormComponent.prototype.valueChanges;
    /**
     * @type {?}
     * @private
     */
    BaseFormComponent.prototype.formCreated;
}
