// dialog-service/esm5/lib/MdcDialogService.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { MdcDialog } from '@angular-mdc/web';
import { Injectable } from '@angular/core';
import { throwError } from 'rxjs';
import { concatMap, finalize, map, tap } from 'rxjs/operators';
import { DialogService } from './DialogService';
import { ProgressComponent } from './mdc/Progress.component';
import { AlertComponent } from './mdc/Alert.component';
import { ConfirmComponent } from './mdc/Confirm.component';
import * as i0 from "@angular/core";
import * as i1 from "@angular-mdc/web/dialog";
/**
 * @deprecated No longer supported. Not enough time to maintain this implementation, contributions
 * welcomed.
 */
var MdcDialogService = /** @class */ (function (_super) {
    tslib_1.__extends(MdcDialogService, _super);
    function MdcDialogService(dialog) {
        var _this = _super.call(this) || this;
        _this.dialog = dialog;
        return _this;
    }
    /**
     * @template T
     * @param {?} work
     * @param {?=} title
     * @return {?}
     */
    MdcDialogService.prototype.withProgress = /**
     * @template T
     * @param {?} work
     * @param {?=} title
     * @return {?}
     */
    function (work, title) {
        title = title || 'Please Wait...';
        /** @type {?} */
        var workOutput;
        /** @type {?} */
        var ref = this.dialog.open(ProgressComponent, {
            clickOutsideToClose: false,
            escapeToClose: false,
            data: { title: title }
        });
        ref.afterOpened().pipe(concatMap((/**
         * @return {?}
         */
        function () { return work; })), tap((/**
         * @param {?} val
         * @return {?}
         */
        function (val) { return workOutput = val; })), // set workOutput
        finalize((/**
         * @return {?}
         */
        function () { return ref.close(); }))).subscribe();
        return ref.afterClosed().pipe(map((/**
         * @return {?}
         */
        function () { return workOutput; })) // return workOutput
        );
    };
    /**
     * @param {?} title
     * @param {?=} options
     * @return {?}
     */
    MdcDialogService.prototype.withAlert = /**
     * @param {?} title
     * @param {?=} options
     * @return {?}
     */
    function (title, options) {
        options = options || {};
        options.acceptButton = options.acceptButton || 'OK';
        /** @type {?} */
        var dialogRef = this.dialog.open(AlertComponent, {
            escapeToClose: false,
            clickOutsideToClose: false,
            data: tslib_1.__assign({ title: title }, options)
        });
        return dialogRef.afterClosed().pipe(map((/**
         * @param {?} result
         * @return {?}
         */
        function (result) { return result === 'accept'; })));
    };
    /**
     * @param {?=} title
     * @param {?=} options
     * @return {?}
     */
    MdcDialogService.prototype.withConfirm = /**
     * @param {?=} title
     * @param {?=} options
     * @return {?}
     */
    function (title, options) {
        title = title || 'Confirm?';
        options = options || {};
        options.acceptButton = options.acceptButton || 'Yes';
        options.cancelButton = options.cancelButton || 'No';
        /** @type {?} */
        var ref = this.dialog.open(ConfirmComponent, {
            escapeToClose: false,
            clickOutsideToClose: false,
            data: tslib_1.__assign({ title: title }, options)
        });
        return ref.afterClosed().pipe(map((/**
         * @param {?} result
         * @return {?}
         */
        function (result) { return result === 'accept'; })));
    };
    /**
     * @param {?} title
     * @param {?} fields
     * @param {?=} options
     * @return {?}
     */
    MdcDialogService.prototype.withForm = /**
     * @param {?} title
     * @param {?} fields
     * @param {?=} options
     * @return {?}
     */
    function (title, fields, options) {
        return throwError('withForm() is not supported in MDC mode');
    };
    MdcDialogService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    MdcDialogService.ctorParameters = function () { return [
        { type: MdcDialog }
    ]; };
    /** @nocollapse */ MdcDialogService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function MdcDialogService_Factory() { return new MdcDialogService(i0.ɵɵinject(i1.MdcDialog)); }, token: MdcDialogService, providedIn: "root" });
    return MdcDialogService;
}(DialogService));
export { MdcDialogService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    MdcDialogService.prototype.dialog;
}
