// dialog-service/esm2015/lib/MdcDialogService.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { MdcDialog } from '@angular-mdc/web';
import { Injectable } from '@angular/core';
import { throwError } from 'rxjs';
import { concatMap, finalize, map, tap } from 'rxjs/operators';
import { DialogService } from './DialogService';
import { ProgressComponent } from './mdc/Progress.component';
import { AlertComponent } from './mdc/Alert.component';
import { ConfirmComponent } from './mdc/Confirm.component';
import * as i0 from "@angular/core";
import * as i1 from "@angular-mdc/web/dialog";
/**
 * @deprecated No longer supported. Not enough time to maintain this implementation, contributions
 * welcomed.
 */
export class MdcDialogService extends DialogService {
    /**
     * @param {?} dialog
     */
    constructor(dialog) {
        super();
        this.dialog = dialog;
    }
    /**
     * @template T
     * @param {?} work
     * @param {?=} title
     * @return {?}
     */
    withProgress(work, title) {
        title = title || 'Please Wait...';
        /** @type {?} */
        let workOutput;
        /** @type {?} */
        const ref = this.dialog.open(ProgressComponent, {
            clickOutsideToClose: false,
            escapeToClose: false,
            data: { title }
        });
        ref.afterOpened().pipe(concatMap((/**
         * @return {?}
         */
        () => work)), tap((/**
         * @param {?} val
         * @return {?}
         */
        val => workOutput = val)), // set workOutput
        finalize((/**
         * @return {?}
         */
        () => ref.close()))).subscribe();
        return ref.afterClosed().pipe(map((/**
         * @return {?}
         */
        () => workOutput)) // return workOutput
        );
    }
    /**
     * @param {?} title
     * @param {?=} options
     * @return {?}
     */
    withAlert(title, options) {
        options = options || {};
        options.acceptButton = options.acceptButton || 'OK';
        /** @type {?} */
        const dialogRef = this.dialog.open(AlertComponent, {
            escapeToClose: false,
            clickOutsideToClose: false,
            data: Object.assign({ title }, options)
        });
        return dialogRef.afterClosed().pipe(map((/**
         * @param {?} result
         * @return {?}
         */
        result => result === 'accept')));
    }
    /**
     * @param {?=} title
     * @param {?=} options
     * @return {?}
     */
    withConfirm(title, options) {
        title = title || 'Confirm?';
        options = options || {};
        options.acceptButton = options.acceptButton || 'Yes';
        options.cancelButton = options.cancelButton || 'No';
        /** @type {?} */
        const ref = this.dialog.open(ConfirmComponent, {
            escapeToClose: false,
            clickOutsideToClose: false,
            data: Object.assign({ title }, options)
        });
        return ref.afterClosed().pipe(map((/**
         * @param {?} result
         * @return {?}
         */
        result => result === 'accept')));
    }
    /**
     * @param {?} title
     * @param {?} fields
     * @param {?=} options
     * @return {?}
     */
    withForm(title, fields, options) {
        return throwError('withForm() is not supported in MDC mode');
    }
}
MdcDialogService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
MdcDialogService.ctorParameters = () => [
    { type: MdcDialog }
];
/** @nocollapse */ MdcDialogService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function MdcDialogService_Factory() { return new MdcDialogService(i0.ɵɵinject(i1.MdcDialog)); }, token: MdcDialogService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    MdcDialogService.prototype.dialog;
}
